File: android/app/src/main/kotlin/jp/cloxs/bittra/ble/BleCoordinator.kt
package jp.cloxs.bitra.ble

import android.bluetooth.BluetoothDevice
import android.content.Context
import android.os.Handler
import android.os.Looper

class BleCoordinator(private val context: Context) {

    // Event callbacks to Flutter
    var onEncounterEvent: ((Map<String, Any>) -> Unit)? = null
    var onStatusEvent: ((Map<String, Any>) -> Unit)? = null
    var onBodyEvent: ((Map<String, Any>) -> Unit)? = null
    var onLogEvent: ((Map<String, Any>) -> Unit)? = null

    private var myTeaser: String = ""
    private var myBody: String = ""

    private val peripheral by lazy { BlePeripheral(context, this::addLog) }
    private val central by lazy { BleCentral(context, this::addLog) }

    // Use MAC address as peerId for Android
    private val peripheralCache = mutableMapOf<String, BluetoothDevice>()

    private val mainHandler = Handler(Looper.getMainLooper())

    init {
        central.onEncounter = { device, senderId, teaser, rssi ->
            if (teaser.isNotEmpty()) {
                mainHandler.post {
                    peripheralCache[device.address] = device
                    onEncounterEvent?.invoke(
                        mapOf(
                            "type" to "encounter",
                            "peerId" to device.address,
                            "senderId" to senderId,
                            "teaser" to teaser,
                            "rssi" to rssi
                        )
                    )
                }
            }
        }
    }

    fun setMyTeaser(teaser: String) {
        myTeaser = PayloadCodec.normalizeTeaser(teaser)
        peripheral.setContent(myTeaser, myBody)
        addLog("APP", "set teaser=$myTeaser")
    }

    fun setMyBody(body: String) {
        myBody = body
        peripheral.setContent(myTeaser, myBody)
        addLog("APP", "set body bytes=${body.toByteArray(Charsets.UTF_8).size}")
    }

    fun startVenueMode() {
        if (myTeaser.isEmpty()) {
            startReceiveOnly()
            return
        }
        peripheral.setContent(myTeaser, myBody)
        peripheral.start()
        central.startScan()
        addLog("APP", "venue ON")
    }

    fun startReceiveOnly() {
        peripheral.stop()
        central.startScan()
        addLog("APP", "receive-only ON")
    }

    fun stopVenueMode() {
        central.stopScan()
        peripheral.stop()
        addLog("APP", "venue OFF")
    }

    fun requestBody(peerId: String) {
        val p = peripheralCache[peerId]
        if (p == null) {
            addLog("APP", "no cached peripheral (wait a bit and try again)")
            onStatusEvent?.invoke(
                mapOf(
                    "type" to "status",
                    "peerId" to peerId,
                    "status" to RequestStatus.FAILED.rawValue,
                    "error" to "Peripheral not cached yet."
                )
            )
            return
        }

        addLog("APP", "requestBody peerId=${p.address}")

        val handler = Handler(Looper.getMainLooper())

        central.onStatus = { st, err ->
            handler.post {
                onStatusEvent?.invoke(
                    mapOf(
                        "type" to "status",
                        "peerId" to peerId,
                        "status" to st.rawValue,
                        "error" to (err ?: "")
                    )
                )
            }
        }

        central.onPreview = { preview ->
            // In Flutter MVP, preview is optional to display progressively
            handler.post {
                onBodyEvent?.invoke(
                    mapOf(
                        "type" to "body",
                        "peerId" to peerId,
                        "preview" to preview
                    )
                )
            }
        }

        central.onBody = { body ->
            handler.post {
                onBodyEvent?.invoke(
                    mapOf(
                        "type" to "body",
                        "peerId" to peerId,
                        "preview" to if (body.length > 80) body.substring(0, 80) else body,
                        "body" to body
                    )
                )
            }
        }

        central.requestBody(p)
    }

    private fun addLog(tag: String, msg: String) {
        android.util.Log.d("BittraBLE", "[$tag] $msg")
        Handler(Looper.getMainLooper()).post {
            onLogEvent?.invoke(
                mapOf(
                    "type" to "log",
                    "tag" to tag,
                    "message" to msg
                )
            )
        }
    }
}

File: ios/Runner/BLECoordinator.swift
import Foundation
import CoreBluetooth

@MainActor
final class BLECoordinator: NSObject {

  var onEncounterEvent: (([String: Any]) -> Void)?
  var onStatusEvent: (([String: Any]) -> Void)?
  var onBodyEvent: (([String: Any]) -> Void)?
  var onDeliveryEvent: (([String: Any]) -> Void)?
  var onLogEvent: (([String: Any]) -> Void)?

  var myTeaser: String = ""
  var myBody: String = ""

  private lazy var peripheral = BLEPeripheral(
    log: self.addLog,
    onDeliveryReceipt: { [weak self] noticeId, count in
      self?.onDeliveryEvent?([
        "type": "delivery",
        "noticeId": noticeId,
        "count": count
      ])
    }
  )
  private lazy var central = BLECentral(log: self.addLog)

  private var peripheralCache: [UUID: CBPeripheral] = [:]

  override init() {
    super.init()

    central.onEncounter = { [weak self] peripheral, senderId, teaser, rssi in
      guard let self = self else { return }
      guard !teaser.isEmpty else { return }
      self.peripheralCache[peripheral.identifier] = peripheral
      self.onEncounterEvent?([
        "type": "encounter",
        "peerId": peripheral.identifier.uuidString,
        "senderId": Int(senderId),
        "teaser": teaser,
        "rssi": rssi
      ])
    }

    central.onStatus = { [weak self] _, _ in
      guard self != nil else { return }
    }

    central.onBody = { [weak self] _ in
      guard self != nil else { return }
    }
  }

  func setMyTeaser(_ teaser: String) {
    myTeaser = PayloadCodec.normalizeTeaser(teaser)
    addLog("APP", "set teaser=\(myTeaser)")
  }

  func setMyBody(_ body: String) {
    myBody = body
    addLog("APP", "set body bytes=\(Data(myBody.utf8).count)")
  }

  func startVenueMode(noticeId: String) {
    precondition(!noticeId.isEmpty, "noticeId must not be empty")
    guard !myTeaser.isEmpty else {
      startReceiveOnly()
      return
    }
    peripheral.setContent(noticeId: noticeId, teaser: myTeaser, body: myBody)
    peripheral.start()
    central.startScan()
    addLog("APP", "venue ON noticeId=\(noticeId)")
  }

  func startReceiveOnly() {
    peripheral.stop()
    central.startScan()
    addLog("APP", "receive-only ON")
  }

  func stopVenueMode() {
    central.stopScan()
    peripheral.stop()
    addLog("APP", "venue OFF")
  }

  func requestBody(forPeerId peerIdString: String) {
    guard let uuid = UUID(uuidString: peerIdString), let p = peripheralCache[uuid] else {
      addLog("APP", "no cached peripheral (wait a bit and try again)")
      onStatusEvent?([
        "type": "status",
        "peerId": peerIdString,
        "status": RequestStatus.failed.rawValue,
        "error": "Peripheral not cached yet."
      ])
      return
    }

    addLog("APP", "requestBody peerId=\(p.identifier)")

    central.onStatus = { [weak self] st, err in
      self?.onStatusEvent?([
        "type": "status",
        "peerId": peerIdString,
        "status": st.rawValue,
        "error": err ?? ""
      ])
    }

    central.onPreview = { [weak self] preview in
      self?.onBodyEvent?([
        "type": "body",
        "peerId": peerIdString,
        "preview": preview
      ])
    }

    central.onBody = { [weak self] body in
      self?.onBodyEvent?([
        "type": "body",
        "peerId": peerIdString,
        "preview": String(body.prefix(80)),
        "body": body
      ])
    }

    central.requestBody(peripheral: p)
  }

  func addLog(_ tag: String, _ msg: String) {
    print("[BITTRA-BLE][\(tag)] \(msg)")
    onLogEvent?([
      "type": "log",
      "tag": tag,
      "message": msg
    ])
  }
}
